sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTargetScaler.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.WebJobs.Host.Scale;
+using Microsoft.Azure.WebJobs.ServiceBus;
+using Microsoft.Azure.WebJobs.ServiceBus.Listeners;
+using System;
+using System.Globalization;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Extensions.ServiceBus.Listeners
+{
+    internal class ServiceBusTargetScaler : ITargetScaler
+    {
+        private readonly string _functionId;
+        private readonly string _entityPath;
+        private readonly ServiceBusMetricsReceiver _serviceBusMetricsReceiver;
+        private readonly ServiceBusOptions _options;
+        private readonly bool _isSessionsEnabled;
+        private readonly TargetScalerDescriptor _targetScalerDescriptor;
+
+        public ServiceBusTargetScaler(string functionId, string entityPath, ServiceBusMetricsReceiver serviceBusMetricsReceiver, ServiceBusOptions options, bool isSessionsEnabled)
+        {
+            _functionId = functionId;
+            _entityPath = entityPath;
+            _serviceBusMetricsReceiver = serviceBusMetricsReceiver;
+            _options = options;
+            _isSessionsEnabled = isSessionsEnabled;
+            _targetScalerDescriptor = new TargetScalerDescriptor($"{_functionId}-ServiceBusTrigger-{_entityPath}".ToLower(CultureInfo.InvariantCulture), functionId);
+        }
+
+        public TargetScalerDescriptor TargetScalerDescriptor => _targetScalerDescriptor;
+
+        public async Task<TargetScalerResult> GetScaleResultAsync(TargetScaleStatusContext context)
+        {
+            ServiceBusTriggerMetrics metrics = await _serviceBusMetricsReceiver.GetMetricsAsync().ConfigureAwait(false);
```

Comment:
GetMetricsAsync does additional work that a target scaler doesn't need, for example peeking messages to get QueueTime. That's an unnecessary performance hit for target scale scenarios. To avoid this, I think we can just have the "metrics receiver" expose a new GetMessageCountAsync() method that does the isListeningOnDeadLetterQueue logic and returns the count, and the GetQueueMetrics/GetTopicMetrics method can be refactored to call that, and you can then just call that from this scaler as well?

Reply:
The additional work is getting partition information - we can add a parameter to existing GetMetricsAsync to skip this. Introducing the new method will require a lot of the code duplication.